Protocols will notify when dhcpcd can exit

rsmarples opened the pull request:
DHCPv6 RELEASE requires the addresses to be dropped before a RELEASE message is sent.
We now wait for an acknowledgement or a timeout before notifying that DHCPv6 has stopped for the interface.

DHCPv4 RELEASE is the other way around, there is no acknowledgement. So we wait for 1 second after sending the message before removing the address and notifying DHCP has stopped for the interface.

If we are not releasing then we notify dhcpcd that the protocol has stopped right away when we drop the lease.

dhcpcd will exit once there are no running protocols for the interfaces.

Fixes #513.
Hopefully #535, #519 and #509 as well.

rsmarples committed:
Protocols will notify when dhcpcd can exit

DHCPv6 RELEASE requires the addresses to be dropped before
a RELEASE message is sent.
We now wait for an acknowledgement or a timeout before notifying
that DHCPv6 has stopped for the interface.

DHCPv4 RELEASE is the other way around, there is no acknowledgement.
So we wait for 1 second after sending the message before removing
the address and notifying DHCP has stopped for the interface.

If we are not releasing then we notify dhcpcd that the protocol has
stopped right away when we drop the lease.

dhcpcd will exit once there are no running protocols for the
interfaces.

Fixes #513.
Hopefully #535, #519 and #509 as well.

rsmarples committed:
Minor improvements.

rsmarples committed:
Fix some fallout.

rsmarples committed:
Run and document RELEASE.

Sime-Zupanovic committed:
Check context options if no interface options are setup.
If allows an interface to stop even if dhcpcd is not.

Sime-Zupanovic commented:
any comment on dhcpcd_handlecarrier() ifp->options->options check?

rsmarples committed:
DHCP6: release cleanly

While here address some hangup issues.

rsmarples committed:
IPv4LL: free on nodrop

Sime-Zupanovic commented:
I see you delivered this patch:
https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/5738d9cbdd7c217094cdb427b91f44892a42acbc

as pasted before, I think this will crash as ctx is wild.

I did just:

--------------------------------- src/dhcpcd.c ---------------------------------
index 5045c0f..d5fe2d9 100755
@@ -775,7 +775,7 @@ dhcpcd_handlecarrier(struct interface *ifp, int carrier, unsigned int flags)
 	ifp->carrier = carrier;
 	ifp->flags = flags;
 
-	if (ifp->options->options & DHCPCD_STOPPING)
+	if (ifp->options && (ifp->options->options & DHCPCD_STOPPING))
 		return;
 
 	if (!if_is_link_up(ifp)) {

Also we need to guard in  dhcp_discover(void *arg) as state was null.

But I will try your additional patches anyway.

rsmarples commented:
> as pasted before, I think this will crash as ctx is wild.

That is not possible as ctx is never removed from the ifp and it's set as the first thing after calloc.

ColinMcInnes commented:
From the logs, there's still some issues with the exit.

```
sudo systemctl stop dhcpcd@eth0; sudo systemctl stop dhcpcd; sleep 5; sudo systemctl start dhcpcd; sudo systemctl start dhcpcd@eth0

Nov 03 11:25:30 systemd[1]: Stopping A minimalistic network configuration daemon with DHCPv4, rdisc and DHCPv6 support...
Nov 03 11:25:30 dhcpcd[26638]: sending signal ALRM to pid 7851
Nov 03 11:25:30 dhcpcd[26638]: waiting for pid 7851 to exit
Nov 03 11:25:30 dhcpcd[26638]: main: eloop_start: Invalid argument
Nov 03 11:25:30 dhcpcd[26638]: sending signal ALRM to pid 7851
Nov 03 11:25:30 dhcpcd[7852]: received SIGALRM, releasing
Nov 03 11:25:30 dhcpcd[7852]: eth0: removing interface
Nov 03 11:25:30 systemd[1]: dhcpcd.service: Control process exited, code=exited, status=1/FAILURE
Nov 03 11:25:30 dhcpcd[26638]: waiting for pid 7851 to exit
Nov 03 11:25:30 dhcpcd[26638]: main: eloop_start: Invalid argument
Nov 03 11:25:30 dhcpcd[7852]: received SIGTERM, stopping
```

rsmarples committed:
eloop: support epoll_pwait2 for kernels >= 5.11

As it more matches ppoll semantics.
Note that epoll_pwait does NOT work with zero events, so
use ppoll in this case.

rsmarples commented on src/eloop.c:
```diff
@@ -50,8 +50,13 @@
 #define USE_EPOLL
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)
+#define HAVE_EPOLL_PWAIT2
+#endif
```
> > If you compile using newer headers than the kernel you are using then you are dumb.
> 
> Not really. Distributions do compile using headers from the latest release, while the running kernel itself often is older. This being said, I cannot think of a distribution that would build using headers from an older major Linux release. For instance, all distributions out there currently ship with Linux 6.x.x kernels so, at worst, the headers are only a few minor releases newer.

I am aware of this, Arch do it all the time and it causes untold problems.
It is a dumb decision because it relies on more complex userland code to deal with ENOSYS.
It takes exactly zero effort for me to ensure the NetBSD kernel I am building and intend to run matches the headers I build software with.

ColinMcInnes replied:
My project compiles against and runs on kernel 5.4.

with 3134f25d1b3b65631fe61d2b4599ebd31e0b7d52 it's now unable to start.

```
Jan 01 00:01:16 dhcpcd[4043]: sandbox: seccomp
Jan 01 00:01:16 dhcpcd[4043]: main: eloop_start: Function not implemented
```

ColinMcInnes replied:
Turns out my kernel headers are 5.11 or newer. They should not be. What a fun surprise, haha.

rsmarples replied:
This is entirely my point.
If I compile against libfoo-2.1.1 then I expect libfoo-2.1.1 features because there is likely no alternative in libfoo-1.x.x without a fair chunk of work.
I just don't see why the kernel is anyway more special here.

eloop is a performance critical section - we want to avoid a kernel context call that does ENOSYS.
Yes we could cache the result, but this is even more complexity.
ok, DHCP is hardly performance critical but the point stands.

rsmarples replied:
Or to put it another way, if too many people complain about this I will rip out epoll_wait2 and just accept the linux path is slow(er) because it has to comptute a timeout from a timespec struct on each entry.

rsmarples committed:
privsep: test eloop error correctly

rsmarples commented:
How are we feeling about merging this to the master branch? Anyone tested it and/or have any objections?

rsmarples committed:
eloop: work with old kernel vs new kernel headers

Because, IMHO, linux distro maintainers are dumb.
There is no logical reason to do this.

rsmarples committed:
eloop: move label

rsmarples commented:
> @rsmarples isn’t new header old kernel image is the user setup issue? The problem is kernel headers and kernel image are 2 independent packages. When one changes kernel he supposed to update them all together, there is even meta package that ties them all together, and that what one should be changing. But there is no obligation to do that, and no clear documentation of which one to use, so you always start with changing wrong one. I have created and fixed this problem locally more time then I can count. Bottom line you do not need to handle this case in dhcpcd, user needs to install proper packages on local system.

in the ideal world i would agree but rn there are two reports of it failing because of this and this isn't even released yet.

rsmarples committed:
Merge branch 'master' into release

rsmarples committed:
Fix exiting

rsmarples committed:
dhcpcd: If globally exiting, ignore carrier

rsmarples committed:
dhcpcd: set exiting on the control socket

Sime-Zupanovic commented:
Can you help me confirm, I understood your https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/d41c505981e1e8f493c544dff004a203fde42973#diff-66945e1e112a14eb5cb5931e2d600af419ed1f4c9a1567669f3de19d0f104c94

1. --exit sets STOPPING.

2.  Each protocol sees STOPPING → calls *_release() → triggers *_deconfigure() → reason="RELEASE".

3. e.g. DHCPv6 RELEASE should be sent to network to signal removal of lease and we should receive ACK on dhcpcd

4. dhcp_free() and dhcpcd_dropped() run in correct order.

5. When all interfaces drop, the main loop exits cleanly.

If we don't have RELEASE ACK timeout callbacks (which also run from eloop) such as
dhcp_discover, dhcp_start, dhcp_bind, dhcp_renew (dhcp.c)
or
dhcp6_* callbacks: dhcp6_start, dhcp6_recvif, dhcp6_freedrop, dhcp6_finishrelease
will be triggered?
I think that's how I saw mentioned dhcp_discover() crash

Program terminated with signal SIGSEGV, Segmentation fault.
#0  dhcp_discover (arg=0x5589e589a0) at dhcp.c:2006
2006	dhcp.c: No such file or directory.
(gdb) bt
#0  dhcp_discover (arg=0x5589e589a0) at dhcp.c:2006
#1  0x00000055757cce64 in eloop_start (eloop=0x5589e47830, signals=signals@entry=0x7fdf70e068) at eloop.c:1201
#2  0x00000055757c5ab4 in main (argc=<optimized out>, argv=<optimized out>, envp=<optimized out>) at dhcpcd.c:2707

(gdb) bt full
#0  dhcp_discover (arg=0x5589e589a0) at dhcp.c:2006
        ifp = 0x5589e589a0
        state = 0x0
        ifo = 0x5589e58bc0
#1  0x00000055757cce64 in eloop_start (eloop=0x5589e47830, signals=signals@entry=0x7fdf70e068) at eloop.c:1201
        error = <optimized out>
        t = 0x5589e47250
        ts = {tv_sec = 0, tv_nsec = 763219470}
        tsp = <optimized out>
        __PRETTY_FUNCTION__ = "eloop_start"

I captured tcpdump during dhcp --exit, but I didn't saw DHCPv6 RELEASE?
On which interface it would be sent?

I appended as well:
https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/6c47ee771cabb1f0597bbf331426c2c3570b238f
https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/7b15c54ddaa5c3b2672452ceff093770ee088a1f
https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/ae22cf0599fcd2899474b5f676b48f8730199e52
https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/93305c28505601fda572912d97a9396b341f8909
https://github.com/NetworkConfiguration/dhcpcd/pull/536/commits/7ef3ba171c855f57df4626eee345f6a257da2d51

Please help.
